### ISSUE TITLE
[Question] Adding an include directory to a WrapDB dependency

### ISSUE THREAD

**Issue by curldivergence** [mesonbuild/meson#12866]
Hi, I have the following problem: I want to use ImGUI from WrapDB and configure it through the provided `imconfig.h`. But the edit I want to make to it (precisely, I want to #define `IMGUI_IMPL_METAL_CPP` there) enables a new requirement for a set of headers that are bundled into my project (but not into ImGUI itself).
So could you advise me, please, what is the cleanest and/or easiest way to impose another include directory to this dependency? As I understand, it can be solved by maintaining my own wrap file, but it seems like an overkill to me since I'm just enabling a feature that is already present in ImGUI's sources.
Thanks!

**Comment by dcbaker** [comment]
I don't think we have support for editing the imconfig.h from our wrap, but you could do something like:
```meson
dep = dependency('imgui', default_options : ['metal=enabled', 'cpp_args='-DIMGUI_IMPL_METAL_CPP'])
```
which would pass the define via the command line, this should work fine. The headers question I'm not sure whether you mean that imgui itself needs to have access to your headers, or if consumers of imgui need them. the latter is achievable like:
```meson
dep = declare_dependency(
  dependencies :  dependency('imgui', default_options : ['metal=enabled', 'cpp_args='-DIMGUI_IMPL_METAL_CPP']),
  include_directories('metal_includes'),
)
```
It would be a sandbox violation to attempt to inject your own headers into the subproject (ie, imgui needs your headers to compile), in that case you'd have to maintain your own wrap.

**Comment by curldivergence** [comment]
Thanks a lot for your reply and explanation!
Unfortunately, I meant the former case (ImGUI needs the headers it’s not bundled with, Apple distributes them in zip archives from the `metal-cpp` doc page, so I bundle them with my project.
And after your reply I see now that I really do need to maintain my wrap, so I’ll go do that :) Thank you!

**Comment by dcbaker** [comment]
Does imgui always need those headers to build metal support? If that's the case and they're not installed by default we could maintain a wrap for that and modify imgui to use that wrap. I think we do that for Vulkan and OpenGL

**Comment by curldivergence** [comment]
No, by default on macOS ImGUI uses the regular Metal framework in ObjC that is installed in a system location (not sure if its preinstalled or is installed with Xcode). But a couple years ago Apple released a set of pure C++ headers (known as `metal-cpp`) that allow to interact with Metal from C++ projects without touching ObjC. And ImGUI added the support of this approach, too, but it’s gated by a disabled by default `ifdef` (since most users don’t need it), and they don’t bundle the C++ wrappers, they are expected to be provided externally.

**Comment by dcbaker** [comment]
I did some reading on this, and I honestly don't see why we couldn't provide a wrap for metal-cpp and and option to our wrap for imgui to support using metal-cpp instead of the objc bindings

**Comment by curldivergence** [comment]
I see two problems in this (so far):
- Apple provides several archives with the C++ wrappers for different macOS versions. Would the WrapDB maintain only the latest version of the wrappers, or all of then? In the latter case, would Meson figure out the correct version automatically, or it would be a user responsibility to set it through options?
- the projects that would include ImGUI with metal-cpp enabled, obviously, would like to utilize these wrappers by themselves, not only to supply them to ImGUI. So when the project source file would include `<Metal/Metal.hpp>`, would it be the same header that ImGUI from WrapDB would bundle, or another copy, bundled by the user’s project in another way? The former seems kinda weird to me and in some cases too restrictive, and the latter opens a possibility for ODR violations if for some reason the headers differ

**Comment by dcbaker** [comment]
1. We generally handle this through "trampoline" wraps. a wrap that basically just checks some condition, and then trampolines to the actual implementation
2. Ideally the project would use the wrap of metal-cpp. copying code into your project is a bad idea, and ODR is one of the reasons why. If you use metal-cpp, and imgui uses metal-cpp, and you both use the wrap they you're ensuring you get the same version every time, and both projects will be recompiled if they change

**Comment by curldivergence** [comment]
I see, thanks for explaining this, your ideas sound reasonable to me. Could I help you in implementing this in some way?
